narwhal/primary/src/state_handler.rs

```diff
@@ -202,45 +237,45 @@ impl RandomnessState {
     }
 
     fn add_message(&mut self, msg: dkg::Message<PkG, EncG>) {
-        if self.used_messages.is_some() {
+        if self.store.has_used_messages() {
             // We've already sent a `Confirmation`, so we can't add any more messages.
             return;
         }
         match self.party.process_message(msg, &mut rand::thread_rng()) {
-            Ok(processed) => {
-                self.processed_messages.push(processed);
-            }
+            Ok(processed) => self
+                .store
+                .add_processed_message(processed.message.sender, processed),
             Err(err) => {
                 debug!("random beacon: error while processing DKG Message: {err:?}");
             }
         }
     }
 
     fn add_confirmation(&mut self, conf: dkg::Confirmation<EncG>) {
-        if self.used_messages.is_none() {
+        if !self.store.has_used_messages() {
             // We should never see a `Confirmation` before we've sent our `Message` because
             // DKG messages are processed in consensus order.
             return;
         }
-        if self.dkg_output.is_some() {
+        if self.has_dkg_output() {
             // Once we have completed DKG, no more `Confirmation`s are needed.
             return;
         }
-        self.confirmations.push(conf)
+        self.store.add_confirmation(conf.sender, conf)
     }
 
     // Generates the next SystemMessage needed to advance the random beacon DKG protocol, if
     // possible, and sends it to the proposer.
     async fn advance_dkg(&mut self) {
         // Once we have enough ProcessedMessages, send a Confirmation.
-        if self.used_messages.is_none() && !self.processed_messages.is_empty() {
-            match self.party.merge(&self.processed_messages) {
+        if !self.store.has_used_messages() {
+            match self.party.merge(&self.store.processed_messages()) {
                 Ok((conf, used_msgs)) => {
                     info!(
                         "random beacon: sending DKG Confirmation with {} complaints",
                         conf.complaints.len()
                     );
-                    self.used_messages = Some(used_msgs);
+                    self.store.set_used_messages(used_msgs);
```

Comment:
If the node crashes after the db write but before sending out the confirmation, does the confirmation need to be resent after restarting? Maybe send the confirmation when `!self.has_dkg_output()` per process lifetime?